Raw Body Parser not works correctly for large file

If the raw body parser exceed the memory limit it will create temporary file, but this not work correctly.
When I run parse.raw in my application for big request I got:
```text
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[InstantiationException: play.api.libs.Files$DefaultTemporaryFileCreator]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:265) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:191) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3.applyOrElse(PlayDefaultUpstreamHandler.scala:273) [play-netty-server_2.11-2.4.3.jar:2.4.3]
	at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$3.applyOrElse(PlayDefaultUpstreamHandler.scala:269) [play-netty-server_2.11-2.4.3.jar:2.4.3]
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:344) [scala-library-2.11.7.jar:na]
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:343) [scala-library-2.11.7.jar:na]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32) [scala-library-2.11.7.jar:na]
	at play.api.libs.iteratee.Execution$trampoline$.executeScheduled(Execution.scala:109) [play-iteratees_2.11-2.4.3.jar:2.4.3]
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:71) [play-iteratees_2.11-2.4.3.jar:2.4.3]
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40) [scala-library-2.11.7.jar:na]
Caused by: java.lang.InstantiationException: play.api.libs.Files$DefaultTemporaryFileCreator
	at java.lang.Class.newInstance(Class.java:427) ~[na:1.8.0_60]
	at play.api.inject.NewInstanceInjector$.instanceOf(Injector.scala:49) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.inject.SimpleInjector$$anonfun$instanceOf$1.apply(Injector.scala:85) ~[play_2.11-2.4.3.jar:2.4.3]
	at scala.collection.MapLike$class.getOrElse(MapLike.scala:128) ~[scala-library-2.11.7.jar:na]
	at scala.collection.AbstractMap.getOrElse(Map.scala:59) ~[scala-library-2.11.7.jar:na]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:85) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:80) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.utils.InlineCache.fresh(InlineCache.scala:69) ~[play_2.11-2.4.3.jar:2.4.3]
Caused by: java.lang.NoSuchMethodException: play.api.libs.Files$DefaultTemporaryFileCreator.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082) ~[na:1.8.0_60]
	at java.lang.Class.newInstance(Class.java:412) ~[na:1.8.0_60]
	at play.api.inject.NewInstanceInjector$.instanceOf(Injector.scala:49) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.inject.SimpleInjector$$anonfun$instanceOf$1.apply(Injector.scala:85) ~[play_2.11-2.4.3.jar:2.4.3]
	at scala.collection.MapLike$class.getOrElse(MapLike.scala:128) ~[scala-library-2.11.7.jar:na]
	at scala.collection.AbstractMap.getOrElse(Map.scala:59) ~[scala-library-2.11.7.jar:na]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:85) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.inject.SimpleInjector.instanceOf(Injector.scala:80) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11-2.4.3.jar:2.4.3]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11-2.4.3.jar:2.4.3]
```

The error start from  backToTemporaryFile() in RawBuffer push method.

Scala: 2.11
Play version: 2.4.3
OS: Windows
The application using macwire not guice...

--- a/framework/project/Build.scala
+++ b/framework/project/Build.scala
@@ -245,6 +245,10 @@ object PlayBuild extends Build {
     .settings(
       addScalaModules(scalaParserCombinators),
       libraryDependencies ++= runtime(scalaVersion.value) ++ scalacheckDependencies,
+      binaryIssueFilters := Seq(
+        // Needed so compile-time DI can use temporary files, raw request bodies
+        ProblemFilters.exclude[MissingMethodProblem]("play.api.BuiltInComponents.tempFileCreator")
+      ),
 
       sourceGenerators in Compile <+= (version, scalaVersion, sbtVersion, sourceManaged in Compile) map PlayVersion,
 

--- a/framework/src/play/src/main/scala/play/api/Application.scala
+++ b/framework/src/play/src/main/scala/play/api/Application.scala
@@ -8,6 +8,7 @@ import javax.inject.Inject
 import akka.actor.ActorSystem
 import com.google.inject.Singleton
 import play.api.http._
+import play.api.libs.Files.{ DefaultTemporaryFileCreator, TemporaryFileCreator }
 import play.api.mvc.EssentialFilter
 import play.api.routing.Router
 import play.api.inject.{ SimpleInjector, NewInstanceInjector, Injector, DefaultApplicationLifecycle }
@@ -266,7 +267,7 @@ trait BuiltInComponents {
 
   def router: Router
 
-  lazy val injector: Injector = new SimpleInjector(NewInstanceInjector) + router + crypto + httpConfiguration
+  lazy val injector: Injector = new SimpleInjector(NewInstanceInjector) + router + crypto + httpConfiguration + tempFileCreator
 
   lazy val httpConfiguration: HttpConfiguration = HttpConfiguration.fromConfiguration(configuration)
   lazy val httpRequestHandler: HttpRequestHandler = new DefaultHttpRequestHandler(router, httpErrorHandler, httpConfiguration, httpFilters: _*)
@@ -282,4 +283,6 @@ trait BuiltInComponents {
 
   lazy val cryptoConfig: CryptoConfig = new CryptoConfigParser(environment, configuration).get
   lazy val crypto: Crypto = new Crypto(cryptoConfig)
+
+  lazy val tempFileCreator: TemporaryFileCreator = new DefaultTemporaryFileCreator(applicationLifecycle)
 }

--- a/framework/src/play/src/main/scala/play/api/inject/BuiltinModule.scala
+++ b/framework/src/play/src/main/scala/play/api/inject/BuiltinModule.scala
@@ -7,6 +7,7 @@ import akka.actor.ActorSystem
 import javax.inject.{ Singleton, Inject, Provider }
 import play.api._
 import play.api.http._
+import play.api.libs.Files.{ DefaultTemporaryFileCreator, TemporaryFileCreator, TemporaryFile }
 import play.api.libs.{ CryptoConfig, Crypto, CryptoConfigParser }
 import play.api.libs.concurrent.{ ExecutionContextProvider, ActorSystemProvider }
 import play.api.routing.Router
@@ -45,7 +46,8 @@ class BuiltinModule extends Module {
       bind[Plugins].toProvider[PluginsProvider],
 
       bind[CryptoConfig].toProvider[CryptoConfigParser],
-      bind[Crypto].toSelf
+      bind[Crypto].toSelf,
+      bind[TemporaryFileCreator].to[DefaultTemporaryFileCreator]
     ) ++ dynamicBindings(
         HttpErrorHandler.bindingsFromConfiguration,
         HttpFilters.bindingsFromConfiguration,

--- a/framework/src/play/src/main/scala/play/api/libs/Files.scala
+++ b/framework/src/play/src/main/scala/play/api/libs/Files.scala
@@ -9,7 +9,6 @@ import java.nio.file.attribute.BasicFileAttributes
 
 import javax.inject.{ Inject, Singleton };
 
-import com.google.inject.ImplementedBy
 import play.api.{ Application, Play };
 import play.api.inject.ApplicationLifecycle;
 import java.nio.file.{ Files => JFiles }
@@ -26,7 +25,6 @@ object Files {
    * file themselves, but this TemporaryFileCreator implementation may also
    * try to clean up any leaked files, e.g. when the Application or JVM stops.
    */
-  @ImplementedBy(classOf[DefaultTemporaryFileCreator])
   trait TemporaryFileCreator {
     def create(prefix: String, suffix: String): File
   }

--- a/framework/src/play/src/main/scala/play/api/libs/Files.scala
+++ b/framework/src/play/src/main/scala/play/api/libs/Files.scala
@@ -9,6 +9,7 @@ import java.nio.file.attribute.BasicFileAttributes
 
 import javax.inject.{ Inject, Singleton };
 
+import com.google.inject.ImplementedBy
 import play.api.{ Application, Play };
 import play.api.inject.ApplicationLifecycle;
 import java.nio.file.{ Files => JFiles }
@@ -25,6 +26,7 @@ object Files {
    * file themselves, but this TemporaryFileCreator implementation may also
    * try to clean up any leaked files, e.g. when the Application or JVM stops.
    */
+  @ImplementedBy(classOf[DefaultTemporaryFileCreator])
   trait TemporaryFileCreator {
     def create(prefix: String, suffix: String): File
   }
@@ -115,7 +117,7 @@ object Files {
     /**
      * Cache the current Application's TemporaryFileCreator
      */
-    private val creatorCache = Application.instanceCache[DefaultTemporaryFileCreator]
+    private val creatorCache = Application.instanceCache[TemporaryFileCreator]
 
     /**
      * Get the current TemporaryFileCreator - either the injected

--- a/framework/src/play/src/test/scala/play/libs/FilesSpec.scala
+++ b/framework/src/play/src/test/scala/play/libs/FilesSpec.scala
@@ -5,6 +5,9 @@ import java.nio.charset.Charset
 
 import org.specs2.mutable.Specification
 import org.specs2.specification.After
+import play.api.ApplicationLoader.Context
+import play.api.routing.Router
+import play.api._
 import play.api.libs.Files.TemporaryFile
 import play.utils.PlayIO
 
@@ -50,6 +53,28 @@ object FilesSpec extends Specification with After {
 
         val to = TemporaryFile(file).moveTo(destination, replace = false)
         new String(java.nio.file.Files.readAllBytes(to.toPath)) must contain("already exists")
+      }
+
+      "works when using compile time dependency injection" in {
+        val context = ApplicationLoader.createContext(
+          new Environment(new java.io.File("."), ApplicationLoader.getClass.getClassLoader, Mode.Test))
+        val appLoader = new ApplicationLoader {
+          def load(context: Context) = {
+            (new BuiltInComponentsFromContext(context) {
+              lazy val router = Router.empty
+            }).application
+          }
+        }
+        val app = appLoader.load(context)
+        Play.start(app)
+        val tempFile = try {
+          val tempFile = TemporaryFile()
+          tempFile.file.exists must beTrue
+          tempFile.file
+        } finally {
+          Play.stop(app)
+        }
+        tempFile.exists must beFalse
       }
 
     }

